### PR TITLE
added setTargeting code for custom key-value pairs

### PR DESCRIPTION
**Pull request by afsheenb**
## Type of change
- [X] Feature

## Description of change
* Added support for custom key-value pairs in setTargeting;
* Move customData, lotameData and and ozoneData objects from top-level object into a separate ext.ozone object

(Resubmitting this after rebasing, was originally PR #3494 )

### CODE REVIEW

File: modules/ozoneBidAdapter.js [hunk 1]
```diff
@@ -274,4 +338,44 @@ export function getTestQuerystringValue() {
...
+// we will use this instead of having to use in the html page: pbjs.bidderSettings = { ozone: { ... } } - see http://prebid.org/dev-docs/publisher-api-reference.html
+pbjs.onEvent('setTargeting', function(arrData) {
+  utils.logInfo('In the setTargeting event handler. Adding custom k/v to the ad request. _ozoneInternal = ', _ozoneInternal);
+  /*
+  In order to add a custom key/value pair you fundamentally need to do one thing:
+  Set new targeting for this key by calling setTargeting() on each window.googletag.pubads().getSlots() array.
+  */
+  // iterate over the slot objects, adding custom targeting parameters
+  utils.logInfo('going to iterate slots: ', window.googletag.pubads().getSlots());
+  window.googletag.pubads().getSlots().forEach(function (slot) {
```
Comment (mkendall07) [inline modules/ozoneBidAdapter.js]: adapters should not be modifying the DFP slots directly. Please drop this part of the code.

**Comment by askrupert** [comment]
Rationale behind including this in the adaptor was to avoid publishers having to add additional on page code to expose these additional key values (which they have been requesting to be made available) ... 

is there any plans to let adaptors expose any custom KVPs in the future beyond the standard 5 that prebid exposes ?

**Comment by idettman** [comment]
You should be able to add extra targeting in the `interpretResponse` by adding a property named `adserverTargeting` to the bidObjects returned. Example: `bidObject.adserverTargeting={oz_biddername_crid:'value'}`. Do a search in the prebid source for `adserverTargeting` to see how this is gets added to targeting

**Comment by AskRupert-DM** [comment]
Thanks @idettman - I'll get a version submitted without this for now for you guys to review - but keen to solve this as the trying to expose some of the information that I am passing as KVPs is tricky to do on page as some of these are dynamic based on the bid response.

Would the following be acceptable - this would pretty much be all handled by Prebid as requested IMO:

pbjs.onEvent('setTargeting', function(arrData) {

// hardcoded values, just to illustrate the proposed solution:

  arrData.mpu['CUSTOM_KEY'] = 'CUSTOM_VALUE';
  arrData.leaderboard['CUSTOM_KEY'] = 'CUSTOM_VALUE';
  targeting.setTargetingForGPT({'mpu' : arrData.mpu});
  targeting.setTargetingForGPT({'leaderboard' : arrData.leaderboard});
});

The above would be added to our adaptor as you see in the previously submitted adaptor code...

**Comment by afsheenb** [comment]
@idettman The setTargeting code has been refactored in the latest commit, please let us know if it's acceptable.

Cheers,

--afsheen

**Comment by idettman** [comment]
Looks good to me, but I'm just waiting on @mkendall07 to review the change since this is out of my wheelhouse.

**Comment by afsheenb** [comment]
Thanks @idettman - we've refactored the code to use the approved way of adding additional targeting and resubmitted.

**Pull request merged**